drivers/platform/mbed/mbed_gpio.cpp

```diff
@@ -0,0 +1,308 @@
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <mbed.h>
+
+// Platform drivers needs to be C-compatible to work with other drivers
+#ifdef __cplusplus
+extern "C"
+{
+#endif //  _cplusplus
+
+#include "no-os/error.h"
+#include "no-os/gpio.h"
+#include "mbed_gpio.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Obtain the GPIO descriptor.
+ * @param desc[in,out] - The GPIO descriptor.
+ * @param param[in] - The structure that contains the GPIO init parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get(struct gpio_desc **desc,
+		      const struct gpio_init_param *param)
+{
+	struct gpio_desc *gpio_desc;
+	struct mbed_gpio_desc *mbed_gpio_desc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	/* Create gpio description object for the device */
+	gpio_desc = (struct gpio_desc *)calloc(1, sizeof(*gpio_desc));
+	if (!gpio_desc)
+		return -ENOMEM;
+
+	gpio_desc->number = param->number;
+
+	/* Create the gpio extra descriptor object to store extra mbed gpio info */
+	mbed_gpio_desc = (struct mbed_gpio_desc *)calloc(1,
+			 sizeof(*mbed_gpio_desc));
+	if (!mbed_gpio_desc)
+		goto err_mbed_gpio_desc;
+
+	mbed_gpio_desc->direction = GPIO_IN;
+	mbed_gpio_desc->gpio_pin = NULL;
+
+	if (param->extra)
+		mbed_gpio_desc->pin_mode = ((struct mbed_gpio_init_param *)
+					    param->extra)->pin_mode;
+	else
+		mbed_gpio_desc->pin_mode = NULL;
+
+	gpio_desc->extra = mbed_gpio_desc;
+	*desc = gpio_desc;
+
+	return SUCCESS;
+
+err_mbed_gpio_desc:
+	free(gpio_desc);
+
+	return -ENOMEM;
+}
+
+/**
+ * @brief Get the value of an optional GPIO.
+ * @param desc[in,out] - The GPIO descriptor.
+ * @param param[in] - GPIO Initialization parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get_optional(struct gpio_desc **desc,
+			       const struct gpio_init_param *param)
+{
+	if (param == NULL) {
+		*desc = NULL;
+		return SUCCESS;
+	}
+
+	return mbed_gpio_get(desc, param);
+}
+
+/**
+ * @brief Free the resources allocated by mbed_gpio_get().
+ * @param desc[in] - The GPIO descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_remove(struct gpio_desc *desc)
+{
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	/* Free the gpio object */
+	if (((struct mbed_gpio_desc *)(desc->extra))->gpio_pin)
+		free(((struct mbed_gpio_desc *)(desc->extra))->gpio_pin);
```

Comment:
I believe this is allocated with `new` always so we should use `delete` to clean it up. It will have undefined behavior otherwise.

Reply:
gpio_pin can be either digital_in Or digital_out instance. Use of delete needs specific typecast to the object and compiler gives warning if not typecasted, so using free here. Is this fine to use free or should we use delete anyway?

Reply:
I also think this should be done with `delete`, it's the proper way to free resources allocated with `new` (performs something more than `free`)